Docs: Ascent Replay Workflow (#1331)
ECP-WarpX/WarpX: Docs/source/visualization/ascent.rst

              db_name: "cinema_out"

You can view the Cinema Database result by opening :code:`cinema_databases/cinema_out/index.html`.

> Replay
> -------------------------

Reviewer: `.rst` files need those aligned, unfortunately:
```suggestion
Replay
------
```

Author: 1. You want me move all PR#1332 commit files to ./Docs/src/visualization/, and combine PR#1332 to PR#1331?
2. What do you mean "save the input file alongside the regular examples"? where should I put the input file?
3. I'll remove the bsub file
4. please show me how to croo-link the png file.
Thanks!

> `Replay <https://ascent.readthedocs.io/en/latest/Utilities.html#getting-data-for-replay>`_ is a utility that allows the user to ‘replay’ Conduit Blueprint HDF5 files saved by Ascent or exported by VisIt (starting in version 3.0 beta) back into Ascent. Replay enables the user or developer to pick specific time steps and load them into Ascent

Reviewer: Suggestion to explain all terms:
```suggestion
With Ascent/Conduit, one can store the intermediate data files before the rendering step is applied to custom files.
These so-called *Conduit Blueprint HDF5 files can be *replayed*, i.e. rendered without running the simulation again.
VisIt 3.0+ also supports those files.

`Replay <https://ascent.readthedocs.io/en/latest/Utilities.html#getting-data-for-replay>`_ is a utility that allows the user to "replay" a simulation from aforementioned files and rendering them with Ascent.
Replay enables the user or developer to pick specific time steps and load them for Ascent visualization, without running the simulation again.
```

> We'll guide you through the procedure of how to use Replay through a WarpX LWFA example: warpx/tools/Ascent/Replay

Reviewer: ```suggestion
We'll guide you through the procedure of how to use Replay through a WarpX LWFA example: `Tools/Ascent/Replay/`
```

> 1. Get Blueprint Files
> -------------------------

Reviewer: Sub-sub-section: https://thomas-cokelaer.info/tutorials/sphinx/rest_syntax.html#headings
```suggestion
Get Blueprint Files
^^^^^^^^^^^^^^^^^^^
```

To use replay, you first need Conduit Blueprint HDF5 files which could be extracted by replay. The following ascent actions file can be used to extract Conduit Blueprint HDF5 files.

.. code-block:: json

    -
      action: action: "add_extracts"
      extracts:
        e1:
          type: "relay"
          params:
            path: "my_output_file_name"
            protocol: "blueprint/mesh/hdf5"

> inputs_3d is LWFA input deck. The data size is 256X256X1024 and the simulation will run for 5000 cycles. WarpX will call ascent every 200 cyces. submit the following job submit script will extract 25 Buleprint HDF5:

Reviewer: ```suggestion
`inputs_3d` is LWFA inputs file.
The data size is 256x256x1024 and the simulation will run for 5000 cycles.
WarpX will call ascent every 200 cycles. submit the following job submit script will extract 25 Blueprint HDF5:
```

.. code-block:: bash

   #!/bin/bash

   #BSUB -P aph114
   #BSUB -W 00:29
   #BSUB -nnodes 2
   #BSUB -alloc_flags smt4
   #BSUB -J WarpX
   #BSUB -o WarpXo.%J
   #BSUB -e WarpXe.%J

   module load gcc
   module load cuda

   export OMP_NUM_THREADS=1
   jsrun -r 6 -a 1 -g 1 -c 7 -l GPU-CPU -d packed -b rs --smpiargs="-gpu" ./warpx inputs_3d

> 2. example Actions
> -------------------------

Reviewer: ```suggestion
Example Actions
^^^^^^^^^^^^^^^
```

> Images of Ex contour with particles could be obtained by replaying on the follow action:

Reviewer: ```suggestion
A visualization of the electric field component `Ex` with a contour plot and with added particles can be obtained by replaying with the following Ascent Action:
```

              sample_rate: 0.90
          f2:
            type: "clip"
            params:
              topology: particle_electrons # particle data
              multi_plane:
                point1:
                  x: 0.0
                  y: 0.0
                  z: 0.0
                normal1:
                  x: 0.0
                  y: -1.0
                  z: 0.0
                point2:
                  x: 0.0
                  y: 0.0
                  z: 0.0
                normal2:
                  x: -0.7
                  y: -0.7
                  z: 0.0

    -
      action: "add_scenes"
      scenes:
        scene1:
          plots:
            p0:
              type: "pseudocolor"
              field: "particle_electrons_uz"
              pipeline: "sampled_particles"
            p1:
              type: "pseudocolor"
              field: "Ex"
              pipeline: "clipped_volume"
          renders:
            image1:
              bg_color: [1.0, 1.0, 1.0]
              fg_color: [0.0, 0.0, 0.0]
              image_prefix: "test%06d"
              camera:
                azimuth: 20
                elevation: 30
                zoom: 2.5

> There are Ascent Actions Examples `Ascent Actions Examples <https://ascent.readthedocs.io/en/latest/Actions/Examples.html#yaml-examples>`_ for you to play.

Reviewer: ```suggestion
There are more `Ascent Actions examples available <https://ascent.readthedocs.io/en/latest/Actions/Examples.html#yaml-examples>`_ for you to play.
```

> 3. Run Replay
> -------------------------

Reviewer: ```suggestion
Run Replay
^^^^^^^^^^
```

> Replay executables are created in the utilities/replay directory of the Ascent instaation. There are two version of replay: mpi version replay_mpi and serial version replay_ser. What you use depends on the data set. Here we use replay_mpi as example.

Reviewer: ```suggestion
For Ascent Replay, two command line tools are provided in the utilities/replay directory of the Ascent installation.
There are two version of replay: the MPI-parallel version ``replay_mpi`` and a serial version, ``replay_ser``.
Use an MPI-parallel replay with data sets created with MPI-parallel builds of WarpX.
Here we use ``replay_mpi`` as an example.
```

The options for replay are:

--root: specifies Blueprint root file to load
--cycles: specifies a text file containing a list of Blueprint root files to load
--actions: specifies the name of the actions file to use (default: ascent_actions.json)

Example launches:

.. code-block:: bash

  srun -n 8 ./replay_mpi --root=test.cycle_002000.root --actions=ascent_action.yaml
  srun -n 8 ./replay_mpi --cycles=warpx_list.txt --actions=ascent_action.yaml

The cycles files list is a text file containing one root file per line:

.. code-block:: bash

  cat warpx_list.txt
  test.cycle_000200.root
  test.cycle_000400.root
  test.cycle_001000.root
  test.cycle_002000.root

Replay will loop over these files in the order in which they appear in the file.

The following is an image from replay on test.cycle_000400.root with above ascent_action.yaml

> .. figure:: WarpX_LWFA_400.png

Reviewer: I fear this file is too large (nearly 0.5MB) to be commited into WarpX directly. I can show you another location to store it later on.
